### PR TITLE
Remove Websocket code from Alfred

### PR DESCRIPTION
**Pull request by alteut**
## Description

Removing the socket.io server from Alfred. Going forward, only Nexus will handle the deltaStream. That is, it will be the only websocket microservice for the client connections.

### CODE REVIEW

File: server/routerlicious/packages/lambdas/src/nexus/index.ts
Comment (sharmakhushboo) [inline server/routerlicious/packages/lambdas/src/nexus/index.ts:1003]: What about lambda/src/alfred/index.ts file?
We need to remove that as well.

**Comment by sharmakhushboo** [comment]
In Nexus lambda, we do not need connectionCountLogger anymore. Its not needed as we have the new connectionCount metric using socket. Please remove its reference from nexus and also remove the connectionCountLogger class/interface as well.

File: server/routerlicious/packages/lambdas/src/nexus/index.ts
Comment (alteut) [reply server/routerlicious/packages/lambdas/src/nexus/index.ts:1003]: It became the src/nexus/index.ts file

**Comment by sharmakhushboo** [comment]
Need to update localOrdererConnection.ts files as well.

**Comment by sharmakhushboo** [comment]
@znewton , could you please review this PR from client-code point of view. If there are any logs/comments that we should update on client side to reflect this shift to nexus, we should do that.

File: server/routerlicious/packages/routerlicious-base/src/test/alfred/io.spec.ts
Comment (alexvy86) [inline server/routerlicious/packages/routerlicious-base/src/test/alfred/io.spec.ts:66]: If this file is purely websocket-related, maybe move it to a nexus folder instead of an alfred one? Or if the tests are mixed, extract the websockets ones to a new file in a nexus folder.

**Review by alexvy86 — COMMENTED** [review]
One general concern from the FF side here: if alfred stops being the one handling deltaStream connections, doesn't the [Helm chart](https://github.com/microsoft/FluidFramework/tree/main/server/routerlicious/kubernetes/routerlicious/templates) need to be updated so that it now spins up Nexus as well to take over?

**Comment by alteut** [comment]
> One general concern from the FF side here: if alfred stops being the one handling deltaStream connections, doesn't the [Helm chart](https://github.com/microsoft/FluidFramework/tree/main/server/routerlicious/kubernetes/routerlicious/templates) need to be updated so that it now spins up Nexus as well to take over?

This is a good catch. Do you know if there is a pipeline that deploys this or what would be a good way to validate this kind of deployment? Is there anyone using these charts directly?

**Comment by alexvy86** [comment]
> This is a good catch. Do you know if there is a pipeline that deploys this or what would be a good way to validate this kind of deployment? Is there anyone using these charts directly?

Yes, the FF team uses them to deploy an internal instance of routerlicious used by our end-to-end tests. The combination of [chart - routerlicious](https://dev.azure.com/fluidframework/internal/_build?definitionId=82) and [deploy - routerlicious](https://dev.azure.com/fluidframework/internal/_build?definitionId=74) package and deploy the helm charts to an internal AKS cluster that our [end-to-end tests pipeline](https://dev.azure.com/fluidframework/internal/_build?definitionId=56) targets. However, there's two environments in that AKS cluster; we automatically deploy to one that is not used by anything (named "ppe"; it's there mostly just to validate that the deployment doesn't fail and that the containers on their own aren't crashing), but deployment to the one that our e2e tests pipeline actually uses requires a manual approval (which I can do).

To validate that the environment created by the charts works as it should, what we can do is create a branch where the e2e tests pipeline targets the ppe environment, so we can see if they still pass after the changes to the charts. I can prepare that branch and hand it over to you so you can run it as you need to make sure tests are passing at the same rate they do when they target the "prod" (non-ppe) environment in AKS (usually 99+% pass rate). Let's sync offline to coordinate.

**Comment by alexvy86** [comment]
One thing I just ran into, is that this split might require more significant changes on our end to support, e.g. https://github.com/microsoft/FluidFramework/blob/main/packages/test/test-drivers/src/routerliciousTestDriver.ts#L57-L62

**Comment by alteut** [comment]
> One thing I just ran into, is that this split might require more significant changes on our end to support, e.g. https://github.com/microsoft/FluidFramework/blob/main/packages/test/test-drivers/src/routerliciousTestDriver.ts#L57-L62

For this the discoveryResponse should override the parameters declared here. The comment a few lines above seems to say so for the other properties. Nexus relies on deltaStreamUrl being set

**Review by alexvy86 — APPROVED** [review]
One last comment below, after that lgtm. As we discussed offline, once this is merged we can check when it deploys to the test environment ("ppe") in FF's internal AKS cluster and I can run our e2e tests against that one. I have a pipeline run against it that succeeded, so we have a baseline.

**Review by alexvy86 — CHANGES_REQUESTED** [review]
Ok found one more place where we need to make updates, the [configmap in the Helm chart](https://github.com/microsoft/FluidFramework/blob/main/server/routerlicious/kubernetes/routerlicious/templates/fluid-configmap.yaml). All the new configuration we added for nexus [here](https://github.com/microsoft/FluidFramework/blob/main/server/routerlicious/packages/routerlicious/config/config.json#L144) probably needs a corresponding entry in the JSON generated by that configmap. I'll probably have to update a couple of files on an internal repo to provide values for the new properties, but I can do that more effectively once we have the changes in the configmap.

**Comment by znewton** [comment]
Not sure if I missed this in the above discussion, but this change will mean that NGINX re-routing of Alfred socket endpoints to Nexus will be a requirement to ensure that older versions of `routerlicious-driver` can still communicate with R11s. I'm not sure if FF's deployment of R11s has NGINX included; if it does, it needs to be updated with routing rules, and if not, we need to solve for that otherwise compat tests will probably fail

**Review by alexvy86 — APPROVED** [review]

File: server/routerlicious/packages/routerlicious-base/src/test/nexus/io.spec.ts
Comment (sharmakhushboo) [inline server/routerlicious/packages/routerlicious-base/src/test/nexus/io.spec.ts:603]: this should be nexusIO?
Check at other places as well

File: server/routerlicious/kubernetes/routerlicious/values.yaml
Comment (sharmakhushboo) [inline server/routerlicious/kubernetes/routerlicious/values.yaml:64]: Let this remain true(default value) in OSS. In frs we do set it to false.

File: server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts
Comment (sharmakhushboo) [inline server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts:46]: Why do we still need this?

File: server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts
Comment (znewton) [inline server/routerlicious/packages/routerlicious-base/src/alfred/runnerFactory.ts:155]: tokenRevocationManager is being started in runner, so we should still be closing, right?

File: packages/test/test-drivers/src/routerliciousTestDriver.ts
Comment (alexvy86) [inline packages/test/test-drivers/src/routerliciousTestDriver.ts:49]: Lines 60-64 below feel like also need a change, for the case where there is no discoveryEndpoint.  Which makes me realize, we probably started using a new subdomain for nexus? If so, we'll also want to add new DNS entries for the FF test cluster, I can do that once we confirm if it's necessary.

File: packages/tools/webpack-fluid-loader/src/getUrlResolver.ts
Comment (alexvy86) [inline packages/tools/webpack-fluid-loader/src/getUrlResolver.ts:75]: I imagine this might work... the concept of the deltaStreamUrl already exists, right? It's just never been used in tinylicious, probably because it wasn't necessary? I think it might be better to update the return object from `tinyliciousUrls()` to also have it set explicitly, even if it's the same URL as the other three.

File: packages/loader/driver-utils/src/insecureUrlResolver.ts
Comment (alexvy86) [inline packages/loader/driver-utils/src/insecureUrlResolver.ts:31]: Nit: let's put this after storageUrl, so the first one remains the "most general" one. The call sites of the constructor won't immediately complain but let's make sure to reorder them there as appropriate.

File: server/routerlicious/packages/lambdas/src/alfred/README.md
Comment (alexvy86) [inline server/routerlicious/packages/lambdas/src/alfred/README.md:1]: Alfred still does something, right? Just not the websocket handling. I'm not familiar enough with it, but can we write a brief description of its new role/responsibilities?

File: packages/loader/driver-utils/src/insecureUrlResolver.ts
Comment (alteut) [reply packages/loader/driver-utils/src/insecureUrlResolver.ts:31]: Oh ok, thought they were doing some kind of alphabetical order

File: server/routerlicious/packages/lambdas/src/alfred/README.md
Comment (alteut) [reply server/routerlicious/packages/lambdas/src/alfred/README.md:1]: This is now described in server/routerlicious-base/alfred where there is an updated README for Alfred

Comment (alteut) [reply server/routerlicious/packages/lambdas/src/alfred/README.md:1]: The /lambdas part of alfred only handled websockets

**Pull request merged**
